### PR TITLE
Dm gift api

### PR DESCRIPTION
**Pull request by iach526526**
Dm gift api, only have permissions can call. Use Disocrd OAuth2 check role

## Summary by CodeRabbit

- **New Features**
  - Introduced a new route for sending gifts to users, enhancing interactivity.
  - Added dynamic redirection after user login.
  - Introduced configuration options for role management related to sending gifts.

- **Improvements**
  - Updated channel identifiers and member counts for better server structure alignment.
  - Enhanced user data management during OAuth authentication.
  - Strengthened error handling across various functionalities.

- **Bug Fixes**
  - Refined error handling in gift sending and database operations.

- **Database Changes**
  - Renamed tables and columns for consistency and improved clarity.
  - Added new columns to the `gift` table for better tracking of transactions.

### CODE REVIEW

**Comment by Edit-Mr** [comment]
你 serverid 那些是不是蓋過去了？

**Review by Edit-Mr — APPROVED** [review]
先這樣

File: app.py [hunk 1]
```diff
@@ -2,9 +2,10 @@
 import json
 import os
 import random
+from urllib.parse import urlencode
 
 # Third-party imports
-from flask import Flask, redirect, request, session, url_for, render_template, send_from_directory
+from flask import Flask, redirect, request, session, url_for, render_template, send_from_directory,jsonify
```
Comment (winstonsung) [inline app.py:8]: Whitespace after comma "`, `".

File: app.py [hunk 2]
```diff
@@ -28,57 +28,184 @@
 github_client_secret = token_data["github_client_secret"]
 github_redirect_uri = token_data["github_redirect_uri"]
 github_discord_redirect_uri = token_data["github_discord_redirect_uri"]
-
+discord_token=token_data["discord_token"]
+send_gift_role=token_data["send_gift_role"]
```
Comment (winstonsung) [inline app.py:32]: Whitespace before and after equal sign "` = `".

File: app.py [hunk 3]
```diff
@@ -28,57 +28,184 @@
...
+    urlencoded = urlencode(params)
+    print(f"{base_url}?\n\n{urlencoded}")
+    return redirect(f"{base_url}?{urlencoded}")
 
 @app.route("/logout")
 def logout():
     session.pop("user", None)
     return redirect(url_for("profile"))
-
+@app.route("/api/send/<int:target_user_id>")
```
Comment (winstonsung) [inline app.py:58]: Place blank line before `@` decorator for functions/methods.

File: app.py [hunk 4]
```diff
@@ -28,57 +28,184 @@
...
+            end(connect,cursor)
+        except Exception as e:
+            return jsonify({"result":"interal server error(SQL) when insert gift","status":500,"error":str(e)})
+        response = requests.post(url, headers=headers, json=json_data,timeout=10)
+        if response.status_code != 200:
+            return jsonify({"error": "Failed to send message","status":response.status_code})
+        return jsonify({"result":"success","status":200})
+    except Exception as e:
+        return jsonify({"result":"interal server error","status":500,"error":str(e)})
 @app.route("/callback")
```
Comment (winstonsung) [inline app.py:155]: Place blank line before `@` decorator for functions/methods.

File: app.py [hunk 5]
```diff
@@ -28,57 +28,184 @@
...
+            "email": user_data.get("email", "No email provided"),
+            "headers" : headers
+        }
+        urlencoded = urlencode(params)
+        separator = '&' if '?' in redirurl else '?'
+        return redirect(f"https://{redirurl}{separator}{urlencoded}")
+    # 否則，重定向到 profile 頁面
     return redirect(url_for("profile"))
-
 @app.route("/github/discord-callback")
```
Comment (winstonsung) [inline app.py:209]: Place blank line before `@` decorator for functions/methods.

File: app.py [hunk 6]
```diff
@@ -348,3 +475,4 @@ def insert_user(user_id, table, cursor): # 初始化（新增）傳入該ID的
 if __name__ == "__main__":
     app.config['TEMPLATES_AUTO_RELOAD'] = True
     app.run(debug = True)
+
```
Comment (winstonsung) [inline app.py:478]: Place blank line before the end of the file.

File: cog/admin_gift.py [hunk 7]
```diff
@@ -7,8 +7,9 @@
 # Local imports
 from build.build import Build
 from cog.core.sql import link_sql, read, write, end
+from cog.core.sendgift import send_gift_button
 
-class AdminRole(Build):
+class SendGift(Build):
```
Comment (winstonsung) [inline cog/admin_gift.py:12]: Consistency. `admin_gift` but `SendGift` ?

File: cog/core/secret.py [hunk 8]
```diff
@@ -5,15 +5,14 @@
 from dotenv import load_dotenv
 import mysql.connector
 
-load_dotenv(f'{os.getcwd()}/.env')
```
Comment (winstonsung) [inline cog/core/secret.py:8]: Should only place 1 blank line.

File: cog/core/secret.py [hunk 9]
```diff
@@ -5,15 +5,14 @@
...
 
 DB_USER = os.getenv('MYSQL_USER')
 DB_PASSWORD = os.getenv('MYSQL_PASSWORD')
 DB_NAME = os.getenv('MYSQL_DATABASE')
 DB_HOST = os.getenv('HOST')
 DB_PORT = os.getenv('MYSQL_PORT')
-
 def connect():
```
Comment (winstonsung) [inline cog/core/secret.py:14]: There should be 1 blank line befor functions/methods.

File: cog/core/sendgift.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Third-party imports
+import discord
+
+from cog.core.sql import link_sql, end
+class MessageSendError(Exception):#自定義的例外類型
```
Comment (winstonsung) [inline cog/core/sendgift.py:5]: There should be 1 blank line before classes.

Comment (winstonsung) [inline cog/core/sendgift.py:5]: No "Mainland China Traditional Chinese". "自訂" not "自定義".

File: cog/core/sendgift.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    pass
+class DBError(Exception):#自定義的例外類型
```
Comment (winstonsung) [inline cog/core/sendgift.py:7]: There should be 1 blank line before classes.

File: cog/core/sendgift.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    pass
+async def send_gift_button(self, target_user: discord.User, gift_type: str, count: int,sender:int) -> None:
```
Comment (winstonsung) [inline cog/core/sendgift.py:9]: There should be 1 blank line before functions/methods.

File: cog/core/sendgift.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        finally:
+            end(connection, cursor)
+
+    try:
+        await target_user.send(embed=embed)
+        msg = await target_user.send(view=view)
+        await record_db(msg.id, gift_type, count, target_user.name)
+    except discord.Forbidden as exc:
+        raise MessageSendError(f"無法向使用者 {target_user.name} 傳送訊息，可能是因為他們關閉了 DM") from exc
+
```
Comment (winstonsung) [inline cog/core/sendgift.py:35]: There should be 1 blank line at the end of the file.

**Comment by winstonsung** [comment]
I'm tired of these formatting issues. Let's just add https://github.com/psf/black to our CI.

**Pull request merged**